From f232720c3ee209e38b1e5af71c5103dcc42e3f3a Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Mon, 28 Jul 2014 12:16:31 -0400
Subject: [fs] cifs: Fix memory leak in cifs_hardlink()

Message-id: <1406549822-10376-39-git-send-email-sprabhu@redhat.com>
Patchwork-id: 86791
O-Subject: [RHEL7.1 PATCH 38/69] [bz1082049]cifs: Fix memory leak in cifs_hardlink()
Bugzilla: 1082049
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Christian Engelmayer <cengelma@gmx.at>

Fix a potential memory leak in the cifs_hardlink() error handling path.
Detected by Coverity: CID 728510, CID 728511.

Signed-off-by: Christian Engelmayer <cengelma@gmx.at>
Signed-off-by: Steve French <smfrench@gmail.com>
(cherry picked from commit abf9767c823bf57837c2032f21332a6efc38a13e)

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/link.c b/fs/cifs/link.c
index e22927b..a017e9c 100644
--- a/fs/cifs/link.c
+++ b/fs/cifs/link.c
@@ -421,8 +421,10 @@ cifs_hardlink(struct dentry *old_file, struct inode *inode,
       CIFS_MOUNT_MAP_SPECIAL_CHR);
  else {
   server = tcon->ses->server;
-  if (!server->ops->create_hardlink)
-   return -ENOSYS;
+  if (!server->ops->create_hardlink) {
+   rc = -ENOSYS;
+   goto cifs_hl_exit;
+  }
   rc = server->ops->create_hardlink(xid, tcon, from_name, to_name,
         cifs_sb);
   if ((rc == -EIO) || (rc == -EINVAL))
-- 
1.7.1